### PR TITLE
IO: Allow running on tmpfs without Direct IO, regardless of direct_io_required

### PR DESCRIPTION
**Pull request by cb22**
Running on `tmpfs` with production settings and a production binary is useful to get system measurement numbers without disk speed playing a role.

This tries to strike a balance between allowing that flexibility and preventing footguns. If `direct_io_required` is true, the code will panic as usual _except_ on a `tmpfs` file system. `tmpfs` is inherently non-durable, so we give an additional warning about this.

### CODE REVIEW

File: src/stdx.zig [hunk 1]
```diff
@@ -487,3 +487,44 @@ test "stdx.zig: parse_dirty_semver" {
...
+            @as(usize, @bitCast(@as(isize, fd))),
+            @intFromPtr(statfs_buf),
+        );
+    } else {
+        return std.os.linux.syscall2(
+            .fstatfs,
+            @as(usize, @bitCast(@as(isize, fd))),
+            @intFromPtr(statfs_buf),
+        );
+    }
```
Comment (batiati) [inline src/stdx.zig]:
```suggestion
    return std.os.linux.syscall2(
        if (@hasField(std.os.linux.SYS, "fstatfs64")) .fstatfs64 else .fstatfs,
        @as(usize, @bitCast(@as(isize, fd))),
        @intFromPtr(statfs_buf),
    );
```

File: src/io/linux.zig [hunk 2]
```diff
@@ -1002,12 +1003,20 @@ pub const IO = struct {
...
             if (direct_io_supported) {
                 flags |= os.O.DIRECT;
             } else if (!constants.direct_io_required) {
                 log.warn("file system does not support Direct I/O", .{});
+            } else if (constants.direct_io_required and dir_on_tmpfs) {
+                // Special case. Normally we would panic here (see below), but being able to
+                // benchmark production workloads on tmpfs is very useful for removing disk speed
+                // from the equation.
+                log.warn("tmpfs is not durable, and your data will be lost on reboot", .{});
+                log.warn("tmpfs does not support Direct I/O - continuing", .{});
```
Comment (batiati) [inline src/io/linux.zig]: This warning of running on tmpfs seems critical. I think we should move the condition outside `if (constants.direct_io)`, therefore not bothering to check `fs_supports_direct_io` in the case of tmpfs. 

That's to prioritize showing the warning when running with `constants.direct_io_required == false`, or even in the eventuality the kernel [changes mind](https://lore.kernel.org/lkml/45A29EC2.8020502@tmr.com/t/) and decides that tmpfs should support `O_DIRECT`.

**Review by batiati — COMMENTED** [review]
It would be great to trick `benchmark.sh` to have an option to create the tmpfs partition 🚀.

**Comment by cb22** [comment]
> It would be great to trick `benchmark.sh` to have an option to create the tmpfs partition 🚀.

Definitely - it's part of what I want to do when we integrate it fully as a subcommand (`tigerbeetle benchmark ...`).

File: src/stdx.zig (hunk 1 above)
Comment (cb22) [reply src/stdx.zig]: Ah yeah - I was just copying the Zig stdlib style, but it's a simple change so no need to keep it consistent with that internally. (should put up a Zig PR :smile:)

File: src/io/linux.zig (hunk 2 above)
Comment (cb22) [reply src/io/linux.zig]: Good point! I've made the warning about running on `tmpfs` always show, and we assume Direct I/O isn't supported.

**Review by batiati — APPROVED** [review]

**Pull request merged**